nslsii/temperature_controllers/lakeshore.py

```diff
@@ -0,0 +1,126 @@
+        '''
+
+        T = Component(EpicsSignalRO, '}T-I')
+        T_celsius = Component(EpicsSignalRO, '}T:C-I')
+        V = Component(EpicsSignalRO, '}Val:Sens-I')
+        status = Component(EpicsSignalRO, '}T-Sts')
+        name = Component(EpicsSignal, '}T:Name-RB', write_pv='}T:Name-SP')
+        alarm_dict = {'high': (EpicsSignalRO, '}Alrm:High-Sts'),
+                      'low': (EpicsSignalRO, '}Alrm:Low-Sts')}
+        alarm = DynamicDeviceComponent(alarm_dict)
+        T_limit = Component(EpicsSignal, '}T:Lim-RB', write_pv='}T:Lim-SP')
+
+    class _Control(PVPositioner):
+        '''A sub-class used to define each of the temperature control channels.
+
+        Parameters
+        ----------
+        *args: list
+            The args to pass down to ``ophyd.PVPositioner.__init__``
+        **kwargs: dict
+            The kwargs to be passed down to the ``ophyd.PVPositioner.__init__``
+        '''
+        # PVPositioner required attributes
+        setpoint = Component(EpicsSignal, '}T-SP')
+        readback = Component(EpicsSignalRO, '}T-RB')
+        done = Component(EpicsSignalRO, '}Sts:Ramp-Sts')
+        # top level attributes
+        heater_range = Component(EpicsSignal, '}Val:Range-Sel')
+        heater_status = Component(EpicsSignalRO, '}Err:Htr-Sts')
+        mode = Component(EpicsSignal, '}Mode-Sel')
+        enable = Component(EpicsSignal, '}Enbl-Sel')
+        target_channel = Component(EpicsSignal, '}Out-Sel')
+        # ramp attributes
+        ramp_dict = {'enabled': (EpicsSignal, '}Enbl:Ramp-Sel'),
+                     'rate': (EpicsSignal, '}Val:Ramp-RB',
+                              {'write_pv': '}Val:Ramp-SP'})}
+        ramp = DynamicDeviceComponent(ramp_dict)
+        # PID loop parameters
+        pid_dict = {'proportional': (EpicsSignal, '}Gain:P-RB',
+                                     {'write_pv': '}Gain:P-SP'}),
+                    'integral': (EpicsSignal, '}Gain:I-RB',
+                                 {'write_pv': '}Gain:I-SP'}),
+                    'derivative': (EpicsSignal, '}Gain:D-RB',
+                                   {'write_pv': '}Gain:D-SP'})}
+        pid = DynamicDeviceComponent(pid_dict)
+        # output parameters
+        output_dict = {'current': (EpicsSignal, '}Out-I'),
+                       'manual_current': (EpicsSignal, '}Out:Man-RB',
+                                          {'write_pv': '}Out:Man-SP'}),
+                       'max_current': (EpicsSignal, '}Out:MaxI-RB',
+                                       {'write_pv': '}Out:MaxI-SP'}),
+                       'resistance': (EpicsSignalRO, '}Out:R-RB',
+                                      {'write_pv': '}Out:R-SP'})}
+        output = DynamicDeviceComponent(output_dict)
+
+    def __init__(self, *args, temperatures=['A', 'B', 'C', 'D'],
+                 controls=[1, 2, 3, 4], **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.temperature = DynamicDeviceComponent(
```

Comment:
These should be defined at the class level, not in the `__init__`.  This should look something like https://github.com/NSLS-II/ophyd/blob/fe812226a193adab0515a81e39fb5e5c22d2ea0c/ophyd/scaler.py#L24-L60

Because there are not defined at class instantiation time, the logic in https://github.com/NSLS-II/ophyd/blob/e098e7f873c8a564fc4c66d2f512286aba3f51cf/ophyd/device.py#L831-L834 will be triggered and correclty set up.  If you do `obj.temperature` you are going to get back the DDC instance, not and instance of `_Temperature`.   This is essentially:

```python
In [18]: class Foo:
    ...:     def __init__(self):
    ...:         self.a = property(lambda self, x: 5, lambda self, x, v: None)

In [19]: f = Foo()

In [20]: f.a
Out[20]: <property at 0x7f75f494b2c8>

```

which does not do what we want.

Reply:
Thanks for the feedback @tacaswell , I suspected strongly that what I was doing would not work and added it in order to start a conversation !-). 

While I agree that this does not do what we want, the common approach you suggest also does not do what we want. After looking at the various instances of this class that exist on the floor at present I found that the major difference between each instance is that they have different numbers of 'sensor' and 'control' attributes. Some beamlines only use 2 sensors and 1 control, others use all 4 sensors and 2 controls etc, etc. In fact there are a few locations where they use a variant of this model that has 8 sensors and 4 controls in the one unit.

As I think it does not make sense to force everyone to include extra sensors and controls that they don't have, or have nothing connected too, there is no way to come up with a 'standard' that is flexible enough to account for the current use cases unless we have a way of defining the number of sensors and controls attributes at instantiation time. Is there an 'acceptable' way that will work to do this that anyone can think of?

Reply:
have made this a factory function for classes which solves this issue